### PR TITLE
Deal with QUERY_OVER_LIMIT response from google

### LINKED ISSUE #585: Improve Google API limit checking for wwwadmin geocoding
See recommendation on this page -- currently, we are hitting the too many requests per second problem, limiting our ability to update geocoding information in bulk.

"Upon receiving a response with status code OVER_QUERY_LIMIT, your application should determine which usage limit has been exceeded. This can be done by pausing for 2 seconds and resending the same request. If status code is still OVER_QUERY_LIMIT, your application is sending too many requests per day. Otherwise, your application is sending too many requests per second." (https://developers.google.com/maps/documentation/business/articles/usage_limits)

### PR DESCRIPTION
**Pull request by adamrp**
This is currently deployed on webdev.  I successfully geocoded 100 in one go (something that failed previously because requests were being submitted too rapidly).  If someone else wants to test, I think this is ready to ship.

Note that there's a limited number of requests that can be submitted in a 24 hour period, so if you're going to test, please test with something like 100 or 200 geocoding attempts.  Also note that there is a 2-second wait if a geocoding attempt comes back with a QUERY_OVER_LIMIT response, so if you're doing 200 at a time, it _could_ take 500 seconds (although it probably won't)

Fixes #585

### CODE REVIEW

File: python_code/data_access/ag_data_access.py [hunk 1]
```diff
@@ -460,26 +465,50 @@ def getMapMarkers(self):
...
+            # if we're over the query limit, wait 2 seconds and try again,
+            # it may just be that we're submitting requests too fast
             if data.get('status', None) == 'OVER_QUERY_LIMIT':
-                return 'over_limit'
+                num_tries += 1
+                sleep(2)
+            elif data.has_key('results'):
+                success = True
             else:
+                raise Exception("first")
```
Comment (ElDeveloper) [inline python_code/data_access/ag_data_access.py]: Hmmm ... if this exception is raised will the next `return 'unknown_error'`ever be reached?

File: python_code/data_access/ag_data_access.py [hunk 2]
```diff
@@ -460,26 +465,50 @@ def getMapMarkers(self):
...
+        # if we got here without getting an unknown_error or succeeding, then
+        # we are over the request limit for the 24 hour period
+        if not success:
+            return 'over_limit'
+
+        # sanity check the data returned by Google and return the lat/lng
+        if len(data['results']) == 0:
+            return 'no_results'
+        if not data['results'][0].get('geometry', {}).get('location',
+                {}).get('lat', {}):
```
Comment (ElDeveloper) [inline python_code/data_access/ag_data_access.py]: Would you mind aligning this in the horrible way that the PEP8 suggests?

File: python_code/data_access/ag_data_access.py [hunk 3]
```diff
@@ -460,26 +465,50 @@ def getMapMarkers(self):
...
+            raise Exception("second")
```
Comment (ElDeveloper) [inline python_code/data_access/ag_data_access.py]: Same thing as above, also the error messages seem to be very "informative".

File: python_code/data_access/ag_data_access.py [hunk 4]
```diff
@@ -460,26 +465,50 @@ def getMapMarkers(self):
...
+            return 'unknown_error'
+        if not data['results'][0].get('geometry', {}).get('location',
+                {}).get('lng', {}):
```
Comment (ElDeveloper) [inline python_code/data_access/ag_data_access.py]: Would you mind aligning this in the horrible way that the PEP8 suggests?

File: python_code/data_access/ag_data_access.py [hunk 5]
```diff
@@ -460,26 +465,50 @@ def getMapMarkers(self):
...
+            raise Exception("third")
```
Comment (ElDeveloper) [inline python_code/data_access/ag_data_access.py]: :broken_heart:

File: python_code/data_access/ag_data_access.py [hunk 6]
```diff
@@ -12,11 +12,16 @@
...
-import json
 import urllib
+import httplib
 from random import choice
+import json
+from time import sleep
+
+import cx_Oracle
+
+class OverGoogleAPILimitException(Exception):
```
Comment (ElDeveloper) [inline python_code/data_access/ag_data_access.py]: Maybe instead `GoogleAPILimitExceeded`? Not of a lot of importance but you know ... just saying.

File: python_code/data_access/ag_data_access.py (hunk 2 above)
Comment (adamrp) [inline python_code/data_access/ag_data_access.py]: Is that aligned with 'location' on the line above?

File: python_code/data_access/ag_data_access.py (hunk 1 above)
Comment (adamrp) [inline python_code/data_access/ag_data_access.py]: :laughing: woops, those were there for debugging -- thanks for catching

File: python_code/data_access/ag_data_access.py (hunk 6 above)
Comment (adamrp) [inline python_code/data_access/ag_data_access.py]: Sure, no problemo

File: python_code/data_access/ag_data_access.py (hunk 1 above)
Comment (ElDeveloper) [inline python_code/data_access/ag_data_access.py]: Cool

File: python_code/data_access/ag_data_access.py (hunk 6 above)
Comment (ElDeveloper) [inline python_code/data_access/ag_data_access.py]: Thanks!

File: python_code/data_access/ag_data_access.py (hunk 2 above)
Comment (ElDeveloper) [inline python_code/data_access/ag_data_access.py]: I think that's the case but I can see how that would not meet the 80 columns limit. Maybe just aligning it with `'geometry'`?

**Comment by adamrp** [comment]
Cool -- I think this is ready to go.  Can we merge and deploy recent code additions today?

**Comment by teravest** [comment]
I'm cool with that,  I have to take off around 1:30 this afternoon but I'm
free all day before that if you guys want to deploy

**Comment by ElDeveloper** [comment]
Thanks for addressing the changes @adamrp.

**Pull request merged**